### ISSUE TITLE
DB schema not found by broker container

### ISSUE THREAD

**Issue by hekonsek** [eclipse/kapua#274]
Hi,

Kapua broker Docker image built from develop branch cannot locate DB schema. Here is the output reported by our users:

```
[EL Severe]: ejb: 2017-02-08 00:56:49.54--ServerSession(730087041)--Exception [EclipseLink-4002] (Eclipse Persistence Services - 2.6.3.v20160428-59c81c5): org.eclipse.persistence.exceptions.DatabaseException
Internal Exception: org.h2.jdbc.JdbcSQLException: Schema "KAPUADB" not found; SQL statement:
SET SCHEMA kapuadb [90079-193]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.engine.Database.getSchema(Database.java:1747)
	at org.h2.command.dml.Set.update(Set.java:408)
	at org.h2.command.CommandContainer.update(CommandContainer.java:98)
	at org.h2.command.Command.executeUpdate(Command.java:258)
	at org.h2.engine.Engine.openSession(Engine.java:206)
	at org.h2.engine.Engine.createSessionAndValidate(Engine.java:154)
	at org.h2.engine.Engine.createSession(Engine.java:137)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:148)
	at java.lang.Thread.run(Thread.java:745)

Error Code: 90079
00:56:49,543 [ActiveMQ NIO Worker 1] INFO  KapuaSecurityBrokerFilter:475  - @@ error
org.eclipse.kapua.KapuaException: An internal error occurred: org.apache.shiro.ShiroException: Error while find user!.
	at org.eclipse.kapua.KapuaException.internalError(KapuaException.java:155)
	at org.eclipse.kapua.service.authentication.shiro.AuthenticationServiceShiroImpl.login(AuthenticationServiceShiroImpl.java:178)
	at org.eclipse.kapua.broker.core.plugin.KapuaSecurityBrokerFilter.addExternalConnection(KapuaSecurityBrokerFilter.java:319)
	at org.eclipse.kapua.broker.core.plugin.KapuaSecurityBrokerFilter.addConnection(KapuaSecurityBrokerFilter.java:280)
	at org.apache.activemq.broker.BrokerFilter.addConnection(BrokerFilter.java:98)
	at org.apache.activemq.broker.MutableBrokerFilter.addConnection(MutableBrokerFilter.java:103)
	at org.apache.activemq.broker.TransportConnection.processAddConnection(TransportConnection.java:849)
	at org.apache.activemq.command.ConnectionInfo.visit(ConnectionInfo.java:139)
	at org.apache.activemq.broker.TransportConnection.service(TransportConnection.java:333)
	at org.apache.activemq.broker.TransportConnection$1.onCommand(TransportConnection.java:197)
	at org.apache.activemq.transport.MutexTransport.onCommand(MutexTransport.java:45)
	at org.apache.activemq.transport.mqtt.MQTTInactivityMonitor.onCommand(MQTTInactivityMonitor.java:162)
	at org.apache.activemq.transport.mqtt.MQTTTransportFilter.sendToActiveMQ(MQTTTransportFilter.java:106)
	at org.apache.activemq.transport.mqtt.MQTTProtocolConverter.sendToActiveMQ(MQTTProtocolConverter.java:181)
	at org.apache.activemq.transport.mqtt.MQTTProtocolConverter.onMQTTConnect(MQTTProtocolConverter.java:310)
	at org.apache.activemq.transport.mqtt.MQTTProtocolConverter.onMQTTCommand(MQTTProtocolConverter.java:205)
	at org.apache.activemq.transport.mqtt.MQTTTransportFilter.onCommand(MQTTTransportFilter.java:94)
	at org.apache.activemq.transport.TransportSupport.doConsume(TransportSupport.java:83)
	at org.apache.activemq.transport.mqtt.MQTTCodec$1.onFrame(MQTTCodec.java:65)
	at org.apache.activemq.transport.mqtt.MQTTCodec.processCommand(MQTTCodec.java:90)
	at org.apache.activemq.transport.mqtt.MQTTCodec.access$400(MQTTCodec.java:26)
	at org.apache.activemq.transport.mqtt.MQTTCodec$4.parse(MQTTCodec.java:213)
	at org.apache.activemq.transport.mqtt.MQTTCodec$3.parse(MQTTCodec.java:179)
	at org.apache.activemq.transport.mqtt.MQTTCodec$2.parse(MQTTCodec.java:138)
	at org.apache.activemq.transport.mqtt.MQTTCodec.parse(MQTTCodec.java:76)
	at org.apache.activemq.transport.mqtt.MQTTNIOTransport.processBuffer(MQTTNIOTransport.java:132)
	at org.apache.activemq.transport.mqtt.MQTTNIOTransport.serviceRead(MQTTNIOTransport.java:120)
	at org.apache.activemq.transport.mqtt.MQTTNIOTransport.access$000(MQTTNIOTransport.java:43)
	at org.apache.activemq.transport.mqtt.MQTTNIOTransport$1.onSelect(MQTTNIOTransport.java:72)
	at org.apache.activemq.transport.nio.SelectorSelection.onSelect(SelectorSelection.java:98)
	at org.apache.activemq.transport.nio.SelectorWorker$1.run(SelectorWorker.java:118)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.shiro.authc.AuthenticationException: org.apache.shiro.ShiroException: Error while find user!
	at org.eclipse.kapua.service.authentication.shiro.KapuaAuthenticator.doMultiRealmAuthentication(KapuaAuthenticator.java:89)
	at org.apache.shiro.authc.pam.ModularRealmAuthenticator.doAuthenticate(ModularRealmAuthenticator.java:269)
	at org.apache.shiro.authc.AbstractAuthenticator.authenticate(AbstractAuthenticator.java:198)
	at org.apache.shiro.mgt.AuthenticatingSecurityManager.authenticate(AuthenticatingSecurityManager.java:106)
	at org.apache.shiro.mgt.DefaultSecurityManager.login(DefaultSecurityManager.java:270)
	at org.apache.shiro.subject.support.DelegatingSubject.login(DelegatingSubject.java:256)
	at org.eclipse.kapua.service.authentication.shiro.AuthenticationServiceShiroImpl.login(AuthenticationServiceShiroImpl.java:147)
	... 32 more
Caused by: org.apache.shiro.ShiroException: Error while find user!
	at org.eclipse.kapua.service.authentication.shiro.realm.UserPassAuthenticatingRealm.doGetAuthenticationInfo(UserPassAuthenticatingRealm.java:104)
	at org.apache.shiro.realm.AuthenticatingRealm.getAuthenticationInfo(AuthenticatingRealm.java:568)
	at org.eclipse.kapua.service.authentication.shiro.KapuaAuthenticator.doMultiRealmAuthentication(KapuaAuthenticator.java:59)
	... 38 more
Caused by: org.eclipse.kapua.KapuaException: An internal error occurred: Error during Persistence Operation.
	at org.eclipse.kapua.KapuaException.internalError(KapuaException.java:139)
	at org.eclipse.kapua.commons.util.KapuaExceptionUtils.convertPersistenceException(KapuaExceptionUtils.java:116)
	at org.eclipse.kapua.commons.jpa.EntityManagerSession._onResult(EntityManagerSession.java:135)
	at org.eclipse.kapua.commons.jpa.EntityManagerSession.onResult(EntityManagerSession.java:106)
	at org.eclipse.kapua.service.user.internal.UserServiceImpl.findByName(UserServiceImpl.java:183)
	at org.eclipse.kapua.service.authentication.shiro.realm.UserPassAuthenticatingRealm.lambda$doGetAuthenticationInfo$0(UserPassAuthenticatingRealm.java:100)
	at org.eclipse.kapua.service.authentication.shiro.realm.UserPassAuthenticatingRealm$$Lambda$2/806672713.call(Unknown Source)
	at org.eclipse.kapua.commons.security.KapuaSecurityUtils.doPriviledge(KapuaSecurityUtils.java:88)
	at org.eclipse.kapua.service.authentication.shiro.realm.UserPassAuthenticatingRealm.doGetAuthenticationInfo(UserPassAuthenticatingRealm.java:100)
	... 40 more
```

**Comment by hekonsek** [comment]
@ctron What version of OpenShift Origin have you used for your recent changes? For 1.4.1 I see the following error when deploying Kapua into OpenShift.

**Comment by ctron** [comment]
I actually did run into the same issues. Thus I added the `-Dcommons.db.schema=` system property. But I am not sure this has anything to do with OpenShift itself. I think it is the liquibase image, which doesn't set the SCHEMA while Kapua itself requires the SCHEMA. Using the system property triggers Kapua to ignore the schema.

**Comment by ctron** [comment]
Ok .. I did a full re-build of Kapua and published the docker images under my own account (ctron) and gave it another try. I do run into a different problem though and will prepare a fix for this:

```
Caused by: java.lang.NumberFormatException: Zero length string
	at java.lang.Integer.decode(Integer.java:1162)
	at org.h2.util.NetUtils.createSocket(NetUtils.java:98)
	at org.h2.engine.SessionRemote.initTransfer(SessionRemote.java:114)
at org.h2.engine.SessionRemote.connectServer(SessionRemote.java:448) 
```

**Comment by hekonsek** [comment]
Just to be sure, `Caused by: java.lang.NumberFormatException: Zero length string` is something you see after setting schema to empty string? And you see that when broker receives new message?

**Comment by hekonsek** [comment]
@ctron Also what OpenShift version do you use BTW? :)

**Comment by ctron** [comment]
No .. I did use the plain `develop` branch. Did a full `mvn clean install -Pdocker` and afterwards `-Pdocker-push` for `assembly` and `dev-tools`.

Then I am calling the openshift script with `DOCKER_ACCOUNT=ctron ./openshift_setup.sh`.

I am using OpenShift 1.4.1.

**Comment by ctron** [comment]
Ok ... I created PR #276 which should fix both issues. For the broker and the API neither the connection parameters nor the schema was set. For the console the schema was set, but the connection parameters where not set.

This connection parameters are necessary after PR #237 got merged. Since it only took care of plain docker, but not of openshift. Now the openshift setup overrides this and it works in both worlds.
